test/e2e/specs/editor/various/change-detection.spec.js

```diff
@@ -0,0 +1,561 @@
+		page,
+		editor,
+		pageUtils,
+	} ) => {
+		// Open the sidebar block settings.
+		await editor.openDocumentSettingsSidebar();
+
+		// Insert a paragraph.
+		await editor.canvas
+			.getByRole( 'button', { name: 'Add default block' } )
+			.click();
+		await page.keyboard.type( 'Hello, World!' );
+
+		// Save and wait till the post is clean.
+		await Promise.all( [
+			expect(
+				page
+					.getByRole( 'region', { name: 'Editor top bar' } )
+					.getByRole( 'button', { name: 'saved' } )
+			).toBeDisabled(),
+			pageUtils.pressKeys( 'primary+s' ),
+		] );
+
+		// Change the paragraph's `drop cap`.
+		await page
+			.getByRole( 'region', { name: 'Editor settings' } )
+			.getByRole( 'button', { name: 'Typography options' } )
+			.click();
+		await page
+			.getByRole( 'menu', { name: 'Typography options' } )
+			.getByRole( 'menuitemcheckbox', { name: 'Show drop cap' } )
+			.click();
+
+		await page
+			.getByRole( 'region', { name: 'Editor settings' } )
+			.getByRole( 'checkbox', { name: 'Drop cap' } )
+			.setChecked( true );
+
+		// Check that the post is dirty.
+		await expect(
+			page
+				.getByRole( 'region', { name: 'Editor top bar' } )
+				.getByRole( 'button', { name: 'Save draft' } )
+		).toBeEnabled();
+
+		// Save and wait till the post is clean.
+		await Promise.all( [
+			expect(
+				page
+					.getByRole( 'region', { name: 'Editor top bar' } )
+					.getByRole( 'button', { name: 'saved' } )
+			).toBeDisabled(),
+			pageUtils.pressKeys( 'primary+s' ),
+		] );
+
+		// Change the paragraph's `drop cap` again.
+		await page
+			.getByRole( 'region', { name: 'Editor settings' } )
+			.getByRole( 'checkbox', { name: 'Drop cap' } )
+			.setChecked( false );
+
+		// Check that the post is dirty.
+		await expect(
+			page
+				.getByRole( 'region', { name: 'Editor top bar' } )
+				.getByRole( 'button', { name: 'Save draft' } )
+		).toBeEnabled();
+	} );
+} );
+
+class ChangeDetectionUtils {
+	/** @type {Page} */
+	#page;
+	/** @type {(() => void) | null} */
+	#continueInterceptedSave = null;
+
+	constructor( { page } ) {
+		this.#page = page;
+	}
+
+	/**
+	 * @return {Promise<boolean>} Whether the post is dirty.
+	 */
+	getIsDirty = async () => {
+		return await test.step(
+			'assert the post is dirty',
+			async () => {
+				const hasDialog = new Promise( ( resolve ) => {
+					this.#page.on( 'dialog', ( dialog ) => {
+						void dialog.accept();
+						resolve( dialog.type() === 'beforeunload' );
+					} );
+					this.#page.on( 'load', () => resolve( false ) );
+				} );
+				await this.#page.reload( { waitUntil: 'commit' } ); // No need to wait for the full load.
+				return hasDialog;
+			},
+			{ box: true }
+		);
+	};
+
+	interceptSave = async () => {
+		let hadInterceptedSave = false;
+
+		const deferred = new Promise( ( res ) => {
+			this.#continueInterceptedSave = res;
+		} );
+
+		await this.#page.route(
+			( url ) =>
+				POST_URLS.some( ( postUrl ) => url.href.includes( postUrl ) ),
```

Comment:
That pattern matcher wasn't working with `wp-env`; maybe it could work with pretty permalinks, but we can ensure that. I found this pattern to be the safe solution for now.

I usually check in UI Mode if requests were successfully intercepted.

## Screenshot

![CleanShot 2024-02-09 at 11 29 23](https://github.com/WordPress/gutenberg/assets/240569/780da96d-2aae-4323-b449-2fe55b81f539)